📖 Refactor InfraCluster contract (#11213)
kubernetes-sigs/cluster-api: docs/book/src/developer/providers/contracts/infra-cluster.md

# Contract rules for InfraCluster

> Infrastructure providers SHOULD implement an InfraCluster resource.

Author: e.g Kubermark has only machines (and it works in combination with machines)

> The goal of an InfraCluster resource is to supplie whatever prerequisites (in term of infrastructure) are necessary for running machines.

Author: oops! fixed

Examples might include networking, load balancers, firewall rules, and so on.

The InfraCluster resource will be referenced by one of the Cluster API core resources, Cluster.

The [Cluster's controller](../../core/controllers/cluster.md) will be responsible to coordinate operations of the InfraCluster,
and the interaction between the Cluster's controller and the InfraCluster resource is based on the contract
rules defined in this page.

Once contract rules are satisfied by an InfraCluster implementation, other implementation details 
could be addressed according to the specific needs (Cluster API in not prescriptive).

Nevertheless, it is always recommended to take a look at Cluster API controllers,

> in tree providers, other providers and use them as a reference implementation (unless custom solutions are required 

Author: fixed

in order to address very specific needs).

In order to facilitate the initial design for each InfraCluster resource, a few [implementation best practices] and [infrastructure Provider Security Guidance]
are explicitly called out in dedicated pages.

<aside class="note warning">

<h1>Never rely on Cluster API behaviours not defined as a contract rule!</h1>

When developing a provider, you MUST consider any Cluster API behaviour that is not defined by a contract rule

> as a Cluster API internal implementation detail, subject to change without any advice.

Author: rephrased, PTAL (also the next sentence)

> When implementing an Infra Cluster, you MUST NOT rely on any Cluster API internal behaviour, otherwise you are exposing users

Author: fixed

> to the risk that such internal behaviour will change.

Author: rephrased, PTAL

Instead, whenever you need something more from the Cluster API contract, you MUST engage the community.

The Cluster API maintainers welcome feedback and contributions to the contract in order to improve how it's defined,
its clarity and visibility to provider implementers and its suitability across the different kinds of Cluster API providers.

To provide feedback or open a discussion about the provider contract please [open an issue on the Cluster API](https://github.com/kubernetes-sigs/cluster-api/issues/new?assignees=&labels=&template=feature_request.md)
repo or add an item to the agenda in the [Cluster API community meeting](https://git.k8s.io/community/sig-cluster-lifecycle/README.md#cluster-api).

</aside>

## Rules (contract version v1beta1)

| Rule                                                                      | Mandatory | Note                                                                |
|---------------------------------------------------------------------------|-----------|---------------------------------------------------------------------|

> | [All resources scope]                                                     | Yes       |                                                                     |

Reviewer: Some character to highlight the first part would make this easier to read, e.g.

[All resources] scope

Or

All resources: scope

(I prefer the first option)
(Same for the headlines below of course)

Author: [] exists because those are links 😅 
Accordingly, option 1 is not viable, implemented option 2

| [InfraCluster terminal failures]                                          | No        |                                                                     |
| [InfraClusterTemplate, InfraClusterTemplateList resource definition]      | No        | Mandatory for ClusterClasses support                                |
| [Externally managed infrastructure]                                       | No        |                                                                     |
| [Multi tenancy]                                                           | No        | Mandatory for clusterctl CLI support                                |
| [Clusterctl support]                                                      | No        | Mandatory for clusterctl CLI support                                |

Note:
- `All resources` refers to all the provider's resources "core" Cluster API interacts with;
  In the context of this page: `InfraCluster`, `InfraClusterTemplate` and corresponding list types

### All resources scope

All resources MUST be namespace-scoped.

### All resources `TypeMeta` and `ObjectMeta` field

> All resources MUST have the standard Kubernetes “type metadata” and “object metadata”

Author: fixed

### All resources `APIVersion` field value

> All resources `APIVersion` is a combination of API group and version.

Reviewer: this reads a bit strange 
```suggestion
All resources MUST have an `APIVersion` value that is a combination of API group and version.
```

(But it's not like Kubernetes allows you to freely chose the value. It will just force you to use apiGroup/version)

I would drop this one

Author: fixed, PTAL

#### All resources API group

> API group MUST be an API group served by the Kubernetes apiserver.

Reviewer: This is a bit strange. Whatever apiGroup you write in your CRD will be served by the Kubernetes apiserver when you deploy the CRD

Or are we saying the apiGroup you choose must be valid according to Kubernetes validation rules?

(to me this sentence seems a bit unnecessary)

The domain for Cluster API resources is `cluster.x-k8s.io`, and infrastructure providers under the Kubernetes SIGS org
generally use `infrastructure.cluster.x-k8s.io` as API group.

If your provider uses a different API group, you MUST grant full read/write RBAC permissions for resources in your API group
to the Cluster API core controllers. The canonical way to do so is via a `ClusterRole` resource with the [aggregation label]
`cluster.x-k8s.io/aggregate-to-manager: "true"`.

The following is an example ClusterRole for a `FooCluster` resource in the `infrastructure.foo.com` API group:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
    name: capi-foo-clusters
    labels:
      cluster.x-k8s.io/aggregate-to-manager: "true"
rules:
- apiGroups:
    - infrastructure.foo.com
  resources:
    - fooclusters
  verbs:
    - create
    - delete
    - get
    - list
    - patch
    - update
    - watch
```

Note: The write permissions allow the Cluster controller to set owner references and labels on the InfraCluster” resources;

> write permissions are not used for general mutations of InfraCluster resources.

Reviewer: I think with ClusterClass, this last sentence is wrong

I would drop this entire Note

Reviewer: Also we'll need permissions for the InfraClusterTemplate (IIRC for this one it might only be for ownerRefs)

Author: Fixed, PTAL

#### All resources version

The resource Version defines the stability of the API and its backward compatibility guarantees.
Examples include `v1alpha1`, `v1beta1`, `v1`, etc. and are governed by the [Kubernetes API Deprecation Policy].

Your provider MUST abide by the same policies.

> Note: The version of your provider do not need to be in sync with the version of core Cluster API resources.

Reviewer: ```suggestion
Note: The version of your provider does not need to be in sync with the version of core Cluster API resources.
```

Author: fixed

Instead, prefer choosing a version that matches the stability of the provider API and its backward compatibility guarantees.

Additionally:

Providers MUST set `cluster.x-k8s.io/<version>` label on the InfraCluster Custom Resource Definitions.

> The label is a map from an API Version of Cluster API (contract) to your Custom Resource Definition versions.

Reviewer: ```suggestion
The label is a map from a Cluster API contract version to your Custom Resource Definition versions.
```
I think

Author: fixed

the last available version in the list if multiple versions are found.

To apply the label to CRDs it’s possible to use commonLabels in your `kustomize.yaml` file, usually in `config/crd`:

```yaml
commonLabels:
  cluster.x-k8s.io/v1alpha2: v1alpha1
  cluster.x-k8s.io/v1alpha3: v1alpha2
  cluster.x-k8s.io/v1beta1: v1beta1
```

An example of this is in the [Kubeadm Bootstrap provider](https://github.com/kubernetes-sigs/cluster-api/blob/release-1.1/controlplane/kubeadm/config/crd/kustomization.yaml).

### InfraCluster, InfraClusterList resource definition

You MUST define a InfraCluster resource.
The InfraCluster resource name must have the format produced by `sigs.k8s.io/cluster-api/util/contract.CalculateCRDName(Group, Kind)`.

Note: Cluster API is using such a naming convention to avoid an expensive CRD lookup operation when looking for labels from 
the CRD definition of the InfraCluster resource.

It is a generally applied convention to use names in the format `${env}Cluster`, where ${env} is a, possibly short, name
for the environment in question. For example `GCPCluster` is an implementation for the Google Cloud Platform, and `AWSCluster`
is one for Amazon Web Services.

```go

> //+kubebuilder:object:root=true

Reviewer: ```suggestion
// +kubebuilder:object:root=true
```

// +kubebuilder:resource:path=fooclusters,shortName=foocl,scope=Namespaced,categories=cluster-api
// +kubebuilder:storageversion
// +kubebuilder:subresource:status

// FooCluster is the Schema for fooclusters.
type FooCluster struct {
    metav1.TypeMeta `json:",inline"`

>     metav1.ListMeta `json:"metadata,omitempty"`

Reviewer: ```suggestion
    metav1.ListMeta `json:"metadata,omitempty"`

```

Author: fixed, good catch

    Spec FooClusterSpec `json:"spec,omitempty"`
    Status FooClusterStatus `json:"status,omitempty"`
}

type FooClusterSpec struct {
    // See other rules for more details about mandatory/optional fields in InfraCluster spec.
    // Other fields SHOULD be added based on the needs of your provider.
}

type FooClusterStatus struct {
    // See other rules for more details about mandatory/optional fields in InfraCluster status.
    // Other fields SHOULD be added based on the needs of your provider.
}
```

For each InfraCluster resource, you MUST also add the corresponding list resource.
The list resource MUST be named as `<InfraCluster>List`.

```go

> //+kubebuilder:object:root=true

Reviewer: ```suggestion
// +kubebuilder:object:root=true
```

// FooClusterList contains a list of fooclusters.

> type FooaClusterList struct {

Reviewer 2: ```suggestion
type FooClusterList struct {
```

    metav1.TypeMeta `json:",inline"`
    metav1.ListMeta `json:"metadata,omitempty"`
    Items           []FooCluster `json:"items"`
}
```

### InfraCluster control plane endpoint

Each Cluster needs a control plane endpoint to sit in front of control plane machines.
Control plane endpoint can be provided in three ways in Cluster API: by the users, by the control plane provider or
by the infrastructure provider.

In case you are developing an infrastructure provider which is responsible to provide a control plane endpoint for

> each Cluster, the host and port of the generated control plane endpoint SHOULD surface on `spec.controlPlaneEndpoint`

Reviewer: ```suggestion
each Cluster, the host and port of the generated control plane endpoint MUST surface on `spec.controlPlaneEndpoint`
```

Is this not a must if "you are developing an infrastructure provider which is responsible to provide a control plane endpoint for
each Cluster"

Not sure how it could work otherwise

    // +optional
    ControlPlaneEndpoint APIEndpoint `json:"controlPlaneEndpoint"`
    
    // See other rules for more details about mandatory/optional fields in InfraCluster spec.
    // Other fields SHOULD be added based on the needs of your provider.
}

// APIEndpoint represents a reachable Kubernetes API endpoint.
type APIEndpoint struct {
    // The hostname on which the API server is serving.
    Host string `json:"host"`
    
    // The port on which the API server is serving.
    Port int32 `json:"port"`
}
```

Once `spec.controlPlaneEndpoint` is set on the InfraCluster resource and the [InfraCluster initialization completed],
the Cluster controller will bubble up this info in Cluster's `spec.controlPlaneEndpoint`.

If instead you are developing an infrastructure provider which is NOT responsible to provide a control plane endpoint,
the implementer should exit reconciliation until it sees Cluster's `spec.controlPlaneEndpoint` populated.

### InfraCluster failure domains

In case you are developing an infrastructure provider which has a notion of failure domains where machines should be

> placed in, the list of available failure domains SHOULD surface on `spec.failureDomains` in the InfraCluster resource.

Reviewer: ```suggestion
placed in, the list of available failure domains MUST surface on `spec.failureDomains` in the InfraCluster resource.
```
Same, not sure why it should be just "should"

```go
type FooClusterStatus struct {

>     // FailureDomains is a list of failure domain objects synced from the infrastructure provider.

Reviewer: This is status while the text above and below says spec

(also in "InfraCluster Initialization completed" section)

Author: good catch, fixed

If defined, also InfraCluster's `spec.controlPlaneEndpoint` and `spec.failureDomains` will be surfaced on Cluster's
corresponding field at the same time.

<aside class="note warning">

<h1>Heads up! this will change with the v1beta2 contract</h1>

When the v1beta2 contract will be released (tentative Apr 2025), `status.initialization.provisioned` will be used
instead of `status.ready`. However, `status.ready` will be supported until v1beta1 removal (~one year later).

See [Improving status in CAPI resources].

</aside>

### InfraCluster conditions

According to [Kubernetes API Conventions], Conditions provide a standard mechanism for higher-level
status reporting from a controller.

Providers implementers SHOULD implement `status.conditions` for their InfraCluster resource.
In case conditions are implemented, Cluster API condition type MUST be used.

If a condition with type `Ready` exist, such condition will be mirrored in Cluster's `InfrastructureReady` condition.

Please note that the `Ready` condition is expected to surface the status of the InfraCluster during its own entire lifecycle,

> including initial provisioning, the final deletion process, and the period in between this two moments.

Reviewer: ```suggestion
including initial provisioning, the final deletion process, and the period in between these two moments.
```

See [Cluster API condition proposal] for more context.

<aside class="note warning">

<h1>Heads up! this will change with the v1beta2 contract</h1>

When the v1beta2 contract will be released (tentative Apr 2025), Cluster API will start using Kubernetes metav1.Condition
types and fully comply to [Kubernetes API Conventions].

In order to support providers continuing to use legacy Cluster API condition types, providers transitioning to
metav1.Condition or even providers adopting custom condition types, Cluster API will start to accept `Ready` condition that
provides following information:
- `type`
- `status`
- `reason` (if omitted, a default one will be used)
- `message` (optional)

> Other fields will be ignored

Reviewer: Is this true? What about LastTransitionTime?

```
			// NOTE: we are preserving the original transition time (when the underlying condition changed)
			LastTransitionTime: condition.LastTransitionTime,
```

Author: Fixed

See [Improving status in CAPI resources] for more context.

Please note that provider that will continue to use legacy Cluster API condition types MUST carefully take into account

> the implication of this choice which are described both the document above and in notice at the beginning of the [Cluster API condition proposal].

Reviewer: ```suggestion
the implication of this choice which are described both in the document above and in the notice at the beginning of the [Cluster API condition proposal].
```

    // but instead indicate that something is fundamentally wrong with the FooCluster and that it cannot be recovered.
    // +optional
    FailureReason *capierrors.ClusterStatusError `json:"failureReason,omitempty"`
    
    // FailureMessage will be set in the event that there is a terminal problem reconciling the FooCluster
    // and will contain a more verbose string suitable for logging and human consumption.
    //
    // This field should not be set for transitive errors that can be fixed automatically or with manual intervention,
    // but instead indicate that something is fundamentally wrong with the FooCluster and that it cannot be recovered.
    // +optional
    FailureMessage *string `json:"failureMessage,omitempty"`
    
    // See other rules for more details about mandatory/optional fields in InfraCluster status.
    // Other fields SHOULD be added based on the needs of your provider.
}
```

Once `status.failureReason` and `status.failureMessage` are set on the InfraCluster resource, the Cluster "core" controller
will bubble up those info in the corresponding fields in Cluster's `status`.

> Please that once a Cluster is set detect in Cluster's `status`, the only way to recover is to delete and recreate the Cluster

Reviewer: ```suggestion
Please note that once failureReason/failureMessage is set in Cluster's `status`, the only way to recover is to delete and recreate the Cluster
```
Or something like that

(it is a terminal failure).

<aside class="note warning">

<h1>Heads up! this will change with the v1beta2 contract</h1>

When the v1beta2 contract will be released (tentative Apr 2025), support for `status.failureReason` and `status.failureMessage`
will be dropped.

See [Improving status in CAPI resources].

</aside>

### InfraClusterTemplate, InfraClusterTemplateList resource definition

For a given InfraCluster resource, you should also add a corresponding InfraClusterTemplate resources in order to use it in ClusterClasses.
The template resource MUST be named as `<InfraCluster>Template`.

```go
// +kubebuilder:object:root=true
// +kubebuilder:resource:path=fooclustertemplates,scope=Namespaced,categories=cluster-api
// +kubebuilder:storageversion

// FooClusterTemplate is the Schema for the fooclustertemplates API.

> type InfraClusterTemplate struct {

Reviewer 2: ```suggestion
type FooClusterTemplate struct {
```

Author: 😅 
Fixed

    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    Spec FooClusterTemplateSpec `json:"spec,omitempty"`
}

> type FooaClusterTemplateSpec struct {

Reviewer 2: ```suggestion
type FooClusterTemplateSpec struct {
```

Author: fixed

    Template FooClusterTemplateResource `json:"template"`
}

type FooClusterTemplateResource struct {
    // Standard object's metadata.
    // More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
    // +optional
    ObjectMeta clusterv1.ObjectMeta `json:"metadata,omitempty"`
    Spec FooClusterSpec `json:"spec"`
}
```

NOTE: in this example InfraClusterTemplate's `spec.template.spec` embeds `FooClusterSpec` from InfraCluster. This might not always be
the best choice depending of if/how InfraCluster's spec fields applies to many clusters vs only one.

For each InfraClusterTemplate resource, you MUST also add the corresponding list resource.
The list resource MUST be named as `<InfraClusterTemplate>List`.

```go

> //+kubebuilder:object:root=true

Reviewer: ```suggestion
// +kubebuilder:object:root=true
```

> // InfraClusterTemplateList contains a list of InfraClusterTemplates.
> type InfraClusterTemplateList struct {

Reviewer 2: ```suggestion
// FooClusterTemplateList contains a list of FooClusterTemplates.
type FooClusterTemplateList struct {
```

    metav1.TypeMeta `json:",inline"`
    metav1.ListMeta `json:"metadata,omitempty"`

>     Items           []InfraClusterTemplate `json:"items"`

Reviewer 2: ```suggestion
    Items           []FooClusterTemplate `json:"items"`
```

}
```

### Externally managed infrastructure

> In some cases, users might be required (or choose to) managedinfrastructure out of band and run CAPI on top of already

Reviewer: ```suggestion
In some cases, users might be required (or choose to) manage infrastructure out of band and run CAPI on top of already
```

existing infrastructure.

In order to support this use case, the InfraCluster controller SHOULD skip reconciliation of InfraCluster resources with
the `cluster.x-k8s.io/managed-by: "<name-of-system>"` label, and not update the resource or its status in any way.

Please note that when the cluster infrastructure is externally managed, it is responsibility of external management system
to abide to the following contract rules:
- [InfraCluster control plane endpoint]
- [InfraCluster failure domains]
- [InfraCluster initialization completed]
- [InfraCluster terminal failures]

See the [externally managed infrastructure proposal] for more detail about this use case.

### Multi tenancy

Multi tenancy in Cluster API defines the capability of an infrastructure provider to manage different credentials,
each one of them corresponding to an infrastructure tenant.

> See [infrastructure Provider Security Guidance] for consideration about cloud provider credential management.

Reviewer: ```suggestion
See [infrastructure Provider Security Guidance] for considerations about cloud provider credential management.
```

The [Cluster's controller](../../core/controllers/cluster.md) will be responsible to coordinate operations of the InfraCluster,
and the interaction between the Cluster's controller and the InfraCluster resource is based on the contract
rules defined in this page.

Once contract rules are satisfied by an InfraCluster implementation, other implementation details 
could be addressed according to the specific needs (Cluster API in not prescriptive).

Nevertheless, it is always recommended to take a look at Cluster API controllers,
in-tree providers, other providers and use them as a reference implementation (unless custom solutions are required 
in order to address very specific needs).

In order to facilitate the initial design for each InfraCluster resource, a few [implementation best practices] and [infrastructure Provider Security Guidance]
are explicitly called out in dedicated pages.

<aside class="note warning">

<h1>Never rely on Cluster API behaviours not defined as a contract rule!</h1>

When developing a provider, you MUST consider any Cluster API behaviour that is not defined by a contract rule

> as a Cluster API internal implementation detail, and iternal implementation detail can change at any time.

Reviewer: ```suggestion
as a Cluster API internal implementation detail, and internal implementation details can change at any time.
```

the implementer should exit reconciliation until it sees Cluster's `spec.controlPlaneEndpoint` populated.

### InfraCluster: failure domains

In case you are developing an infrastructure provider which has a notion of failure domains where machines should be
placed in, the list of available failure domains MUST surface on `status.failureDomains` in the InfraCluster resource.

```go
type FooClusterStatus struct {
    // FailureDomains is a list of failure domain objects synced from the infrastructure provider.
    FailureDomains clusterv1.FailureDomains `json:"failureDomains,omitempty"`
    
    // See other rules for more details about mandatory/optional fields in InfraCluster status.
    // Other fields SHOULD be added based on the needs of your provider.
}
```

`clusterv1.FailureDomains` is a map, defined as `map[string]FailureDomainSpec`. A unique key must be used for each `FailureDomainSpec`.
`FailureDomainSpec` is defined as:
- `controlPlane bool`: indicates if failure domain is appropriate for running control plane instances.
- `attributes map[string]string`: arbitrary attributes for users to apply to a failure domain.

> Once `spec.failureDomains` is set on the InfraCluster resource and the [InfraCluster initialization completed],

Reviewer: ```suggestion
Once `status.failureDomains` is set on the InfraCluster resource and the [InfraCluster initialization completed],
```
